server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java

```diff
@@ -157,10 +157,10 @@ public void verifyEngineBeforeIndexClosing() throws IllegalStateException {
 
     protected final DirectoryReader wrapReader(DirectoryReader reader,
                                                     Function<DirectoryReader, DirectoryReader> readerWrapperFunction) throws IOException {
-        reader = ElasticsearchDirectoryReader.wrap(reader, engineConfig.getShardId());
         if (engineConfig.getIndexSettings().isSoftDeleteEnabled()) {
             reader = new SoftDeletesDirectoryReaderWrapper(reader, Lucene.SOFT_DELETES_FIELD);
         }
+        reader = ElasticsearchDirectoryReader.wrap(reader, engineConfig.getShardId());
```

Comment:
Should we wrap the reader as `ElasticsearchDirectoryReader` after applying `readerWrapperFunction`?

Reply:
either way is fine. I tried to apply it at low as possible.

Reply:
I moved the es directory reader on top to ensure that we don't hit this in the future if we add new wrappers in the `readerWrapperFunction`.